src/fftvis/cpu/utils.py

```diff
@@ -22,3 +22,60 @@ def inplace_rot(rot: np.ndarray, b: np.ndarray):  # pragma: no cover
         out[1] = rot[1, 0] * b[0, n] + rot[1, 1] * b[1, n] + rot[1, 2] * b[2, n]
         out[2] = rot[2, 0] * b[0, n] + rot[2, 1] * b[1, n] + rot[2, 2] * b[2, n]
         b[:, n] = out
+
+def prepare_source_catalog(sky_model: np.ndarray, polarized_beam: bool) -> tuple[np.ndarray, bool]:
+    """
+    Prepare the source catalog for the given sky model by building its coherency matrix.
+
+    Parameters
+    ----------
+    sky_model : np.ndarray
+        - Unpolarized: shape (nsources, nfreqs)
+        - Polarized:   shape (nsources, nfreqs, 4)
+    polarized_beam : bool
+        If True, you may pass either a 2D unpolarized sky_model or
+        a 3D polarized array with last axis length 4.
+        If False, only 2D unpolarized sky_models are allowed.
+
+    Returns
+    -------
+    coherency : np.ndarray
+        - If input was 2D: shape (nsources, nfreqs) (monopole)
+        - If input was 3D: shape (nsources, nfreqs, 2, 2) (full coherency matrix)
```

Comment:
This is 4D right?

Reply:
Oh I see you're actually expecting it to be `(nsources, nfreqs, 4)`

Reply:
Yeah, that's right.